### DISCUSSION TITLE
WebRTC Alexa SDP Offer not working: MediaMtx error: "payload type not found"

### DISCUSSION THREAD

**Discussion by JoeHogan** [bluenviron/mediamtx#2539] [Questions]
### Question

I'm new to Alexa Skill development, and MediaMtx, and WebRTC, so that should set the stage for my question... I'm able to get Alexa to initiate a WebRTC Offer for a camera stream, and I have a working MediaMtx WebRTC camera stream, but the Alexa offer response to the POST request is a 400... and MediaMtx log says: 

**media_proxy  | 2023/10/20 21:54:05 INF [WebRTC] [session 63da9321] created by 172.22.0.1:55670
media_proxy  | 2023/10/20 21:54:05 INF [WebRTC] [session 63da9321] closed: payload type not found**

Network connectivity is not a problem since I'm able to simulate this locally by POSTing the Alexa WebRTC SDP Offer (obtained via Lambda logs) to my MediaMtx endpoint, example: "http:192.168.X.X/myCamera/whep" - this results in the 400 response with no returned error message, and the MediaMtx log error above: "payload type not found."

When I instead locally POST the MediaMtx generated Offer (obtained from Chrome dev tools on the front end of the live view web interface), I get a good (200) Answer response.

The Alexa Offer is posted below, as well as the MediaMtx Offer... they are obviously different and result in different outcomes to my POST request. The questions i have are:

1) should i try to make the Alexa Offer something the MediaMtx understands, and if so, is there an Alexa config for this?
2)  should i instead try to make MediaMtx accept what Alexa is Offering, and if so, is there a MediaMtx config for this?
3) What the heck does "payload type not found" mean on MediaMtx WebRTC?

**Alexa Offer:** 

v=0
o=- 3906815152 3906815152 IN IP4 0.0.0.0
s=a 2 z
c=IN IP4 0.0.0.0
t=0 0
a=group:BUNDLE audio0 video0
m=audio 1 UDP/TLS/RTP/SAVPF 96 0 8
a=candidate:1 1 UDP 2013266431 54.196.226.75 57151 typ host
a=candidate:2 1 TCP 1015021823 54.196.226.75 9 typ host tcptype active
a=candidate:3 1 TCP 1010827519 54.196.226.75 48081 typ host tcptype passive
a=candidate:2 2 TCP 1015021822 54.196.226.75 9 typ host tcptype active
a=candidate:3 2 TCP 1010827518 54.196.226.75 52272 typ host tcptype passive
a=candidate:1 2 UDP 2013266430 54.196.226.75 52793 typ host
a=setup:actpass
a=extmap:3 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=rtpmap:96 opus/48000/2
a=rtcp:9 IN IP4 0.0.0.0
a=rtcp-mux
a=sendrecv
a=mid:audio0
a=ssrc:335015742 cname:user2657050833@host-3d8d78c5
a=ice-ufrag:QJiK
a=ice-pwd:HnE3hD5KwNcL/KFJOrNiDV
a=fingerprint:sha-256 60:B0:A0:08:D3:2D:BA:0A:AE:F2:4B:A1:C6:34:AF:9A:F1:FF:98:85:49:30:0A:A9:95:E1:0B:9A:DF:9B:E8:39
m=video 1 UDP/TLS/RTP/SAVPF 99
a=candidate:1 1 UDP 2013266431 54.196.226.75 57151 typ host
a=candidate:2 1 TCP 1015021823 54.196.226.75 9 typ host tcptype active
a=candidate:1 2 UDP 2013266430 54.196.226.75 52793 typ host
a=candidate:3 2 TCP 1010827518 54.196.226.75 52272 typ host tcptype passive
a=candidate:3 1 TCP 1010827519 54.196.226.75 48081 typ host tcptype passive
a=candidate:2 2 TCP 1015021822 54.196.226.75 9 typ host tcptype active
b=AS:2500
a=setup:actpass
a=extmap:3 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=rtpmap:99 H264/90000
a=rtcp:9 IN IP4 0.0.0.0
a=rtcp-mux
a=sendrecv
a=mid:video0
a=rtcp-fb:99 nack
a=rtcp-fb:99 nack pli
a=rtcp-fb:99 ccm fir
a=ssrc:942178419 cname:user2657050833@host-3d8d78c5
a=ice-ufrag:QJiK
a=ice-pwd:HnE3hD5KwNcL/KFJOrNiDV
a=fingerprint:sha-256 60:B0:A0:08:D3:2D:BA:0A:AE:F2:4B:A1:C6:34:AF:9A:F1:FF:98:85:49:30:0A:A9:95:E1:0B:9A:DF:9B:E8:39

**MediaMTX Offer:**

v=0
o=- 4110524935615227415 2 IN IP4 127.0.0.1
s=-
t=0 0
a=group:BUNDLE 0 1
a=extmap-allow-mixed
a=msid-semantic: WMS
m=video 9 UDP/TLS/RTP/SAVPF 96 97 102 103 104 105 106 107 108 109 127 125 39 40 45 46 98 99 100 101 112 113 116 117 118
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:zJWT
a=ice-pwd:iUgAn/PbR55HKCXKJOeDVpWg
a=ice-options:trickle
a=fingerprint:sha-256 BA:E6:42:DF:E8:E2:44:6B:AD:09:1D:EF:80:40:20:B2:C3:C8:88:7B:4B:13:38:BD:DA:4D:B2:BA:D6:D4:48:6D
a=setup:actpass
a=mid:0
a=extmap:1 urn:ietf:params:rtp-hdrext:toffset
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:3 urn:3gpp:video-orientation
a=extmap:4 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:5 http://www.webrtc.org/experiments/rtp-hdrext/playout-delay
a=extmap:6 http://www.webrtc.org/experiments/rtp-hdrext/video-content-type
a=extmap:7 http://www.webrtc.org/experiments/rtp-hdrext/video-timing
a=extmap:8 http://www.webrtc.org/experiments/rtp-hdrext/color-space
a=extmap:9 urn:ietf:params:rtp-hdrext:sdes:mid
a=extmap:10 urn:ietf:params:rtp-hdrext:sdes:rtp-stream-id
a=extmap:11 urn:ietf:params:rtp-hdrext:sdes:repaired-rtp-stream-id
a=sendrecv
a=msid:- 44a485d2-c3d7-4da6-98f2-acc17a33ffc5
a=rtcp-mux
a=rtcp-rsize
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 goog-remb
a=rtcp-fb:96 transport-cc
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtpmap:97 rtx/90000
a=fmtp:97 apt=96
a=rtpmap:102 H264/90000
a=rtcp-fb:102 goog-remb
a=rtcp-fb:102 transport-cc
a=rtcp-fb:102 ccm fir
a=rtcp-fb:102 nack
a=rtcp-fb:102 nack pli
a=fmtp:102 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42001f
a=rtpmap:103 rtx/90000
a=fmtp:103 apt=102
a=rtpmap:104 H264/90000
a=rtcp-fb:104 goog-remb
a=rtcp-fb:104 transport-cc
a=rtcp-fb:104 ccm fir
a=rtcp-fb:104 nack
a=rtcp-fb:104 nack pli
a=fmtp:104 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=42001f
a=rtpmap:105 rtx/90000
a=fmtp:105 apt=104
a=rtpmap:106 H264/90000
a=rtcp-fb:106 goog-remb
a=rtcp-fb:106 transport-cc
a=rtcp-fb:106 ccm fir
a=rtcp-fb:106 nack
a=rtcp-fb:106 nack pli
a=fmtp:106 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=42e01f
a=rtpmap:107 rtx/90000
a=fmtp:107 apt=106
a=rtpmap:108 H264/90000
a=rtcp-fb:108 goog-remb
a=rtcp-fb:108 transport-cc
a=rtcp-fb:108 ccm fir
a=rtcp-fb:108 nack
a=rtcp-fb:108 nack pli
a=fmtp:108 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=42e01f
a=rtpmap:109 rtx/90000
a=fmtp:109 apt=108
a=rtpmap:127 H264/90000
a=rtcp-fb:127 goog-remb
a=rtcp-fb:127 transport-cc
a=rtcp-fb:127 ccm fir
a=rtcp-fb:127 nack
a=rtcp-fb:127 nack pli
a=fmtp:127 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=4d001f
a=rtpmap:125 rtx/90000
a=fmtp:125 apt=127
a=rtpmap:39 H264/90000
a=rtcp-fb:39 goog-remb
a=rtcp-fb:39 transport-cc
a=rtcp-fb:39 ccm fir
a=rtcp-fb:39 nack
a=rtcp-fb:39 nack pli
a=fmtp:39 level-asymmetry-allowed=1;packetization-mode=0;profile-level-id=4d001f
a=rtpmap:40 rtx/90000
a=fmtp:40 apt=39
a=rtpmap:45 AV1/90000
a=rtcp-fb:45 goog-remb
a=rtcp-fb:45 transport-cc
a=rtcp-fb:45 ccm fir
a=rtcp-fb:45 nack
a=rtcp-fb:45 nack pli
a=rtpmap:46 rtx/90000
a=fmtp:46 apt=45
a=rtpmap:98 VP9/90000
a=rtcp-fb:98 goog-remb
a=rtcp-fb:98 transport-cc
a=rtcp-fb:98 ccm fir
a=rtcp-fb:98 nack
a=rtcp-fb:98 nack pli
a=fmtp:98 profile-id=0
a=rtpmap:99 rtx/90000
a=fmtp:99 apt=98
a=rtpmap:100 VP9/90000
a=rtcp-fb:100 goog-remb
a=rtcp-fb:100 transport-cc
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=fmtp:100 profile-id=2
a=rtpmap:101 rtx/90000
a=fmtp:101 apt=100
a=rtpmap:112 H264/90000
a=rtcp-fb:112 goog-remb
a=rtcp-fb:112 transport-cc
a=rtcp-fb:112 ccm fir
a=rtcp-fb:112 nack
a=rtcp-fb:112 nack pli
a=fmtp:112 level-asymmetry-allowed=1;packetization-mode=1;profile-level-id=64001f
a=rtpmap:113 rtx/90000
a=fmtp:113 apt=112
a=rtpmap:116 red/90000
a=rtpmap:117 rtx/90000
a=fmtp:117 apt=116
a=rtpmap:118 ulpfec/90000
a=ssrc-group:FID 1760683812 3991578591
a=ssrc:1760683812 cname:r/q2lLCk+qfSe0pN
a=ssrc:1760683812 msid:- 44a485d2-c3d7-4da6-98f2-acc17a33ffc5
a=ssrc:3991578591 cname:r/q2lLCk+qfSe0pN
a=ssrc:3991578591 msid:- 44a485d2-c3d7-4da6-98f2-acc17a33ffc5
m=audio 9 UDP/TLS/RTP/SAVPF 111 63 9 0 8 13 110 126
c=IN IP4 0.0.0.0
a=rtcp:9 IN IP4 0.0.0.0
a=ice-ufrag:zJWT
a=ice-pwd:iUgAn/PbR55HKCXKJOeDVpWg
a=ice-options:trickle
a=fingerprint:sha-256 BA:E6:42:DF:E8:E2:44:6B:AD:09:1D:EF:80:40:20:B2:C3:C8:88:7B:4B:13:38:BD:DA:4D:B2:BA:D6:D4:48:6D
a=setup:actpass
a=mid:1
a=extmap:14 urn:ietf:params:rtp-hdrext:ssrc-audio-level
a=extmap:2 http://www.webrtc.org/experiments/rtp-hdrext/abs-send-time
a=extmap:4 http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01
a=extmap:9 urn:ietf:params:rtp-hdrext:sdes:mid
a=sendrecv
a=msid:- 3b92e12f-bc2d-439f-a360-fc6f11bb0734
a=rtcp-mux
a=rtpmap:111 opus/48000/2
a=rtcp-fb:111 transport-cc
a=fmtp:111 minptime=10;useinbandfec=1;stereo=1;sprop-stereo=1
a=rtpmap:63 red/48000/2
a=fmtp:63 111/111
a=rtpmap:9 G722/8000
a=rtpmap:0 PCMU/8000
a=rtpmap:8 PCMA/8000
a=rtpmap:13 CN/8000
a=rtpmap:110 telephone-event/48000
a=rtpmap:126 telephone-event/8000
a=ssrc:3948914358 cname:r/q2lLCk+qfSe0pN
a=ssrc:3948914358 msid:- 3b92e12f-bc2d-439f-a360-fc6f11bb0734

**Comment by kakopappa** [comment]
When I worked on Alexa WebRTC integration for Sinric Pro with MediaMTX I noticed the same problem. Issue was related to Alexa Echo SPD offer had some profiles which seems to upset MediaMTX. When I removed them from Alexa SPD it started working. I did complain to Amazon Alexa support about it but didn’t go anywhere.

btw you can use Sinric Pro for 3 cameras for free forever. Don’t need to do Skill setup and Lambda all of that.

tutorial https://help.sinric.pro/pages/tutorials/camera/raspberrypi-webrtc-alexa-googlehome-camera

**Comment by JoeHogan** [comment]
Thanks very much for your response! Not sure what you mean by Echo profiles... as far as I can tell, the Alexa api is agnostic, and I have but one dev account linked to this skill.

I looked at sinric, and nothing against it, but its more of the same in terms of camera limits and possibly timeout limits... would like to free people of their Blink camera 10 minute live streams, especially when the stream is local.

**Comment by kakopappa** [comment]
These are the changes I made to Alexa SDP ofter to get it working for Sinric Pro

1. Replace `m=audio 1 UDP/TLS/RTP/SAVPF 96 0 8` with `m=audio 1 UDP/TLS/RTP/SAVPF 96` 
2. Replace `'a=group:BUNDLE audio0 video0,` with `a=group:BUNDLE video0`
3. Replace `a=group:BUNDLE video0.*m=video,` with `m=video`

Good luck with your project. That sounds like the right thing to do

**Reply by JoeHogan** [reply]
You're the man! I don't know if I can configure Alexa to make a different offer, but maybe I can intercept and modify it. Seems counter to the whole idea of offer/answer, but im willing to try!

**Reply by kakopappa** [reply]
Should be able to do this in lambda.

```
  let offer = event.directive.payload.offer.value;
  offer = offer.replace('a=group:BUNDLE audio0 video0', 'a=group:BUNDLE video0');
  offer = offer.replace('a=group:BUNDLE video0.*m=video', 'm=video');
  offer = offer.replace('96 0 8', '96');
```

**Reply by JoeHogan** [reply]
Excellent! That's more or less what I hope to do in my lambda... ill try to simulate it locally first with a modified Alexa sdp... but not till tomorrow!

**Reply by JoeHogan** [reply]
just tried this locally and it seems to work now, thanks so much! wish i didnt have to manipulate the Alexa offer... not sure how long that will work without a breaking change... if anyone knows a more "configuration friendly" solution, that would be great!

**Comment by aler9** [comment]
This is a bug in the underlying WebRTC library, that requires a `rtpmap` or `fmtp` attribute for each payload format declared in `m=audio`. However, payload types 0 and 8 are static payloads with no available parameters, thus the rtpmap is redundant.

I've created a patch: https://github.com/pion/sdp/pull/165

Discussion is moved there. Next time, please open an issue and not a discussion.
